data/txDupCache.go

```diff
@@ -170,17 +192,17 @@ func (c *txSaltedCache) innerSwap(scheduled bool) {
 
 	if scheduled {
 		// updating by timer, the prev size is a good estimation of a current load => preallocate
-		c.cur = make(map[crypto.Digest]struct{}, len(c.prev))
+		c.cur = make(map[crypto.Digest]cacheValue, len(c.prev))
 	} else {
 		// otherwise start empty
-		c.cur = map[crypto.Digest]struct{}{}
+		c.cur = map[crypto.Digest]cacheValue{}
```

Comment:
The else case here it means it's not scheduled. We should probably pre-allocate maxsize here since it's one of two cases:

1. The node has just started up
2. The cache rotated due to previous cache being full

In case of high traffic we would be adding extra allocation churn here since we would be hitting case 2 often and allocating incrementally. If the network is not at capacity we would free up over-allocated memory after the first scheduled interval

Reply:
but without load it would cause memory waste. Maybe start with 1/3 of the max size or something?

Reply:
But for only a maximum of 1 minute or whatever the timeout is set to be. If we are rotating because of capacity it's likely that the next one will fill up as well and we will have to allocate more memory to get it back up to maximum. 

What's odd to me is that if we rotated on a schedule when the cache was 99% full we will start with 99% of maxsize pre-allocated but if we actually fill it up 100% we will start with it empty.

If we never fill up we will never-allocate full maxsize other than at node startup. 

A compromise would be to start at 1/3 maxsize when starting the node up (or empty) and then re-allocate max if we've expired but it's a bit of a complication for marginal benefits

Reply:
use len(c.prev) unconditionally now